File: lintcode/611.knight-shortest-path.cpp
/**
 * Definition for a point.
 * struct Point {
 *     int x;
 *     int y;
 *     Point() : x(0), y(0) {}
 *     Point(int a, int b) : x(a), y(b) {}
 * };
 */

class Solution {

private:
    vector<vector<int>> directions = {{1, 2}, {1, -2}, {-1, 2}, {-1, -2}, {2, 1}, {2, -1}, {-2, 1}, {-2, -1}};

    bool in_bound(vector<vector<bool>> &grid, int x, int y)
    {
        return  x >= 0 && x < grid.size() && y >= 0 && y < grid[x].size();
    }

public:
    /**
     * @param grid: a chessboard included 0 (false) and 1 (true)
     * @param source: a point
     * @param destination: a point
     * @return: the shortest path 
     */
    int shortestPath(vector<vector<bool>> &grid, Point &source, Point &destination) {
        // write your code here

        queue<pair<int, int>> q;
        q.push(make_pair(source.x, source.y));
        grid[source.x][source.y] = true;

        int steps = 0;

        while(!q.empty())
        {
            int size = q.size();
            for (auto i = 0; i < size; ++i)
            {
                pair<int, int> tmp = q.front();
                q.pop();

                if (tmp.first == destination.x && tmp.second == destination.y)
                {
                    return steps;
                }

                for (auto i = 0; i < 8; i++)
                {
                    int new_x = tmp.first + directions[i][0];
                    int new_y = tmp.second + directions[i][1];
                    if (in_bound(grid, new_x, new_y) && grid[new_x][new_y] == 0)
                    {
                        q.push(make_pair(new_x, new_y));
                        grid[new_x][new_y] = true;
                    }
                }
            }
    
            steps++;
        }

        return -1;
    }
};

File: lintcode/611.knight-shortest-path.py
"""
Definition for a point.
class Point:
    def __init__(self, a=0, b=0):
        self.x = a
        self.y = b
"""

class Solution:
    """
    @param grid: a chessboard included 0 (false) and 1 (true)
    @param source: a point
    @param destination: a point
    @return: the shortest path 
    """
    def shortestPath(self, grid, source, destination):
        # write your code here

        directions = [[1, 2], [1, -2], [-1, 2], [-1, -2], [2, 1], [2, -1], [-2, 1], [-2, -1]]
        q = [(source.x, source.y)]
        steps = 0
        grid[source.x][source.y] = True

        while len(q) > 0:
            size = len(q)

            for i in range(size):

                x, y = q.pop(0)

                if x == destination.x and y == destination.y:
                    return steps

                for i in range(8):
                    new_x = x + directions[i][0]
                    new_y = y + directions[i][1]
                    if self.in_bound(grid, new_x, new_y) and not grid[new_x][new_y]:
                        grid[new_x][new_y] = True
                        q.append((new_x, new_y))

            steps += 1

        return -1

    def in_bound(self, grid, x, y):
        return 0 <= x < len(grid) and 0 <= y < len(grid[x])